docs/proposed/config-split.md

```diff
@@ -0,0 +1,420 @@
+aliases for some of them to align with the naming conventions of the
+repositories.
+
+The layout of options into sections in mbedtls_config.h does not suit
+TF-PSA-Crypto well thus the configuration options tf_psa_crypto_config.h are
+organized into different sections (see below).
+
+## Configuration files and config.py
+
+Each repository contains a config.py script to create and modify configurations.
+
+In Mbed TLS, config.py handles both mbedtls_config.h and
+tf_psa_crypto_config.h. It can set or unset TLS, x509 and cryptographic
+configuration options without having to specify the configuration file the
+options belong to. Commands like full and baremetal affect both configuration
+files.
+
+In TF-PSA-Crypto, config.py addresses only tf_psa_crypto_config.h.
+
+## Sections in tf_psa_crypto_config.h
+
+The tf_psa_crypto_config.h configuration file is organized into eight sections.
+
+The pre-split mbedtls_config.h configuration files contains configuration
+options that apply to the whole code base (TLS, x509, crypto and tests) mostly
+related to the platform abstraction layer and testing. In tf_psa_crypto_config.h
+these configurations options are organized into two sections, one for the
+platform abstraction layer options and one for the others, respectively named
+"Platform abstraction layer" and "General and test configuration options".
+
+Then, the "Cryptographic mechanism selection (PSA API)" section is the
+equivalent of the pre-split crypto_config.h configuration file containing the
+PSA_WANT_ prefixed macros.
+
+The following section named "Cryptographic mechanism selection (extended API)"
+contains the configuration options for the cryptography mechanisms that are not
+yet part of the PSA cryptography API (like LMS or PK).
+
+It is followed by the "Data format support" section that contains configuration
+options of utilities related to various data formats (like base64 or ASN1 APIs).
+These utilities aim to facilitate the usage of the PSA cryptography API in other
+cryptography projects.
+
+Compared to Mbed TLS, the cryptography code in TF-PSA-Crypto is not located
+in a single directory but split between the PSA core (core directory) and the
+PSA builtin drivers (drivers/builtin/src directory). This is reflected in
+tf_psa_crypto_config.h with two sections respectively named "PSA core" and
+"Builtin drivers".
+
+Finally, the last section named "Legacy cryptography" contains the configuration
+options that will eventually be removed as duplicates of PSA_WANT_\* and
+MBEDTLS_PSA_ACCEL_\* configuration options.
+
+By contrast to mbedtls_config.h, tf_psa_crypto_config.h does not contain a
+section like the "Module configuration options" one containing non boolean
+configuration options. The configuration options that are not boolean are
+located in the same section as the boolean option they are associated to.
+
+Open question: do we group them into a subsection?
```

Comment:
> I wonder if the separation between boolean vs non-boolean was for the benefit of config.py

It dates from the PolarSSL days — so you tell me? `config.py` cares about sections because it's backward-compatible with `config.pl`. The non-boolean section is skipped by `config.py full`.

This is something we should probably change. More than once we've accidentally put a boolean option in the wrong section and it didn't get activated by `full`. It'll potentially affect the set of configurations that we're testing, but a new major release is a good time for that.

Reply:
Uncommenting everything with `config.py` full would be a good thing. And on a related note [we mostly aren't testing non-boolean options](https://github.com/Mbed-TLS/mbedtls/issues/4048). From an architectural perspective, I'm all for making sections purely for the sake of documentation, and having them ignored by tooling. However, from a planning perspective, we may have to resolve CI complaints when we do this.

Reply:
Thanks for your comments. I won't have a lot of time to split the file (before the end of September) and address all the consequences of that. Thus I may need to be pragmatic and group the non boolean options into subsections, we will see. I think we have a consensus on keeping non boolean options in the same sections as their corresponding boolean options.